patches/server/0962-Registry-and-Lifecycle-API.patch

```diff
@@ -0,0 +1,1926 @@
++                return (ExtendedRegistry<API>) getOrCreateRegistry(registryKey1, registry);
++            } else if (registryKey instanceof RegistryKey.EnumMutableRegistryKey<? extends RegistryKey.RegistryLayerType, ?> registryKey1) {
++                return (ExtendedRegistry<API>) getOrCreateRegistry((RegistryKey.EnumMutableRegistryKey<? extends RegistryKey.RegistryLayerType, ?>) registryKey1, registry);
++            }
++            throw new IllegalStateException("Illegal Registry Key type: " + registryKey.getClass());
++        }
++
++        /**
++         *  The converter it gets from the map is "strictly-cached",
++         *  meaning that if there is already an api registry instance for that nms registry instance,
++         *  then it gets from the cache rather than creating a new one.
++         */
++        @SuppressWarnings("unchecked")
++        public <API, NMS> ExtendedRegistry<API> getOrCreateRegistry(RegistryKey<? extends RegistryKey.RegistryLayerType, API> registryKey, net.minecraft.core.Registry<NMS> registry) {
++            Converters.Converter<ExtendedRegistry<API>, net.minecraft.core.Registry<NMS>> converter = (Converters.Converter<ExtendedRegistry<API>, net.minecraft.core.Registry<NMS>>) this.keyConverterMap.get(registryKey);
++            ExtendedRegistry<API> apiRegistry = converter.convertNMS(registry);
++            this.apiRegistryPerRegistryKey.put(registryKey, apiRegistry);
++            return apiRegistry;
++        }
++
++        @SuppressWarnings("unchecked")
++        public <API, NMS> WritableRegistry<API> getOrCreateRegistry(RegistryKey.MutableRegistryKey<? extends RegistryKey.RegistryLayerType, API> registryKey, net.minecraft.core.Registry<NMS> registry) {
++            Converters.Converter<ExtendedRegistry<API>, net.minecraft.core.Registry<NMS>> converter = (Converters.Converter<ExtendedRegistry<API>, net.minecraft.core.Registry<NMS>>) this.keyConverterMap.get(registryKey);
++            WritableRegistry<API> apiRegistry = (WritableRegistry<API>) converter.convertNMS(registry);
++            this.apiRegistryPerRegistryKey.put(registryKey, apiRegistry);
++            return apiRegistry;
++        }
++
++        @SuppressWarnings("unchecked")
++        public <API extends Enum<API> & Keyed, NMS> EnumWritableRegistry<API> getOrCreateRegistry(RegistryKey.EnumMutableRegistryKey<? extends RegistryKey.RegistryLayerType, API> registryKey, net.minecraft.core.Registry<NMS> registry) {
++            Converters.Converter<ExtendedRegistry<API>, net.minecraft.core.Registry<NMS>> converter = (Converters.Converter<ExtendedRegistry<API>, net.minecraft.core.Registry<NMS>>) this.keyConverterMap.get(registryKey);
++            EnumWritableRegistry<API> apiRegistry = (EnumWritableRegistry<API>) converter.convertNMS(registry);
++            this.apiRegistryPerRegistryKey.put(registryKey, apiRegistry);
++            return apiRegistry;
++        }
```

Comment:
Regarding this, is this only used in order to dynamically add new enum types?

Reply:
The lines you marked there are used to get or create API EnumWritableRegistry from the NMS one. EnumWritableRegistry is an API interface which is used to specifically register Enums types by users, I made it because Bukkit enums implement the Keyed interface and have NamespacedKey as a field, we don't need to have Key or ResourceKey param in the register method in the WritableRegistry, because the Enum already has the namespaced key when the user created it, so `EnumWritableRegistry#register()` method contains only `E extends Enum<E> & Keyed` as a param, and gets the key from it to register it.